### ISSUE TITLE
Example of JSON/Struct

### ISSUE THREAD

**Issue by knvpk** [Jocoon/php-parquet#1]
I started using this package and we are migrating the data in postgres to s3, all went well but unable to find how to work with json fields (static typed).

While seeing the doc found `StructField` but unable to find how to put data in to it.

when using the StructField im getting below error:

```
 Argument 1 passed to jocoon\parquet\data\DataColumn::__construct() must be an instance of jocoon\parquet\data\DataField, instance of jocoon\parquet\data\StructField given, called in
```

**Comment by knvpk** [comment]
ping @KevinVonJocoon

**Comment by KevinVonJocoon** [comment]
Hey there. Nice to hear your successful production-level usage of this package.
I'd like to refer to https://github.com/Jocoon/php-parquet/blob/master/tests/StructureTest.php for your question, though it uses some helper functions, but it demonstrates the usage of StructFields.

For your specific case, I'd like to propose to rely on pure String-Datafields. Though your JSON might be more-or-less staticly typed, but JSON is not meant to provide you safety regarding this issue.

For a small round-up:
Using structure fields, you'll have to provide your data (in DataColumns) in the same order you created your schema - meaning: a StructField is more-or-less 'virtual' and you have to call WriteColumn in the same order, recursively, using the field elements of the StructField.

'_StructField_' is not a derived class of _DataField_, but simply _Field_ and therefore, is unsupported by DataColumn. But I'm going to think about an alternative/additional way of saving StructField's data by extending DataColumn to overcome this conceptual quirkyness.

Right now, we've shoveled around 400GB of pure data from our production DB to our 'parquet-based archive'. As this package is still beta-y, I've always performed double-checks when using this library, meaning:
When exporting (and especially deleting stuff from your database), re-validate the stored data by querying it again and doing a 1:1 comparison of the stored parquet data and your 'live' data.

I'm closing this issue. Feel free to re-open or open other issues, as far as needed.

**Comment by knvpk** [comment]
Thanks @KevinVonJocoon , Thanks for the information, i didin't looked in to tests previously, after reading it i written a recursive function to keep data in it as below but getting below error

```
        $cols = $this->getColumns();
        $schema = new Schema($cols);

            foreach ($cols as $col) {
                $this->writeColumnRecursive($all_requests_modified, $groupWriter, $col);
            }

private function getColumns()
    {
        return [
            DataField::createFromType('id', 'string'),
            DataField::createFromType('path', 'string'),
            DataField::createFromType('method', 'string'),
            DataField::createFromType('request_id', 'string'),
            DataField::createFromType('query', 'string'),
            DataField::createFromType('request_data', 'string'),
            DataField::createFromType('response_data', 'string'),
            DataField::createFromType('response_code', 'integer'),

            StructField::createWithFieldArray("client_details", [
                StructField::createWithFieldArray("location", [
                    DataField::createFromType("lat", "decimal"),
                    DataField::createFromType("lon", "decimal"),
                ]),
                StructField::createWithFieldArray("device", [
                    DataField::createFromType("model", "string"),
                    DataField::createFromType("type", "string"),
                ]),
                StructField::createWithFieldArray("address", [
                    DataField::createFromType("country", "string"),
                    DataField::createFromType("state", "string"),
                    DataField::createFromType("city", "string"),
                    DataField::createFromType("postal_code", "string"),
                ]),
                DataField::createFromType("timezone", "string"),
                DataField::createFromType("ip", "string"),
            ]),

            // DataField::createFromType("client_details", 'string'),
            DataField::createFromType('source', 'string'),
            DataField::createFromType('platform', 'string'),
            DataField::createFromType('version', 'string'),
            DataField::createFromType('device_id', 'string'),
            DataField::createFromType('user_id', 'string'),
            DateTimeDataField::create('created_at', $this->datetime_format),
            DateTimeDataField::create('updated_at', $this->datetime_format),
        ];
    }


    private function writeColumnRecursive(Collection $data, ParquetRowGroupWriter $groupWriter, Field $col)
    {
        try {

            if ($col instanceof StructField) {
                $fields = $col->getFields();
                foreach ($fields as $field) {
                    $this->writeColumnRecursive($data, $groupWriter, $field);
                }
            } else {
                $data_column = new DataColumn($col, $data->pluck($col->path)->toArray());
                $groupWriter->WriteColumn($data_column);
            }
        } catch (TypeError $ex) {
            dd($col, $col->path, $ex->getMessage());
        }
    }
``` 

Error message:

```
Return value of jocoon\parquet\file\ThriftSchemaTree::FindInternal() must be an instance of jocoon\parquet\file\Node, null returned
```

that error occured while processing column with path `client_details.device.model`

**Comment by KevinVonJocoon** [comment]
Thanks for providing info on this issue (though it would have been better to have some sample data beforehand ;)
I could reproduce your issue, which was simply an incomplete return type which caused this respective TypeError/exception.
This has now been fixed in v0.4.2 (commit 13cc071). CI Tests have already passed, please test and provide feedback, if you like to.

I could also reproduce your processing at this point and would like to recommend to use `$schema->getDataFields();` instead of traversing the schema manually. `Schema::getDataFields()` retrieves all fields (in a linear way) for writing out your recursive structure fields (so you can skip `$col instanceof StructField` checks in your method `writeColumnRecursive`).

**Comment by knvpk** [comment]
Yeah its working now after updating to .2 and using directly getDataFields rather than own traversing, thanks.

I actually write a public get count method on ParquetWriter to get the `count($this->openedWriters)`. this is usefull when i need to know how many groups i have written so far.
